src/main/java/com/google/sps/ChatWebSocket.java

```diff
@@ -59,9 +64,46 @@ public void onOpen(Session session) throws IOException {
     }
 
     @OnMessage
-    public void textMessage(Session session, String msg) {
+    public void textMessage(Session session, String msg) throws IOException {
+        JSONObject messageJson = new JSONObject(msg);
+
+        switch (messageJson.getString("type")) {
+            case "MSG_SEND":
+                chatMessage(session, messageJson);
+                break;
+            case "MAP_SEND":
+
+                break;
+            default:
+                System.out.println("Invalid message");
+        }
+    }
+
+    private void chatMessage(Session session, JSONObject messageData)
+            throws IOException {
+        String msg = messageData.getString("message");
         System.out.println("Text message: " + msg);
+
+        List<Session> participants =
+            WebSocketHandler.getInstance().getRoomList(chatRoom);
+
+        JSONObject echoData = new JSONObject();
+        echoData.put("type", "MSG_RECV");
+        echoData.put("message", messageData.getString("message"));
+        echoData.put("uid", uid);
+        String echoJson = echoData.toString();
+
+        for (Session participant : participants) {
+            participant.getBasicRemote().sendText(echoJson);
+        }
     }
+
+    private void mapMessage(Session session, JSONObject messageData) {
```

Comment:
ditto. Update name to what it does. If this is incomplete, please add a TODO comment with what needs to be done